Toolkit/ArcGISToolkit/TemplatePickerViewController.swift

```diff
@@ -0,0 +1,296 @@
+        
+        // load map, get initial data
+        self.map?.load(){ [weak self] error in
+            
+            guard let map = self?.map else{
+                return
+            }
+            
+            if error == nil{
+                self?.getInitialData(map: map)
+            }
+        }
+    }
+    
+    private func makeSearchController() -> UISearchController {
+        let searchController = UISearchController(searchResultsController: nil)
+        searchController.dimsBackgroundDuringPresentation = false
+        searchController.hidesNavigationBarDuringPresentation = false
+        searchController.searchResultsUpdater = self
+        
+        let searchBar = searchController.searchBar
+        searchBar.spellCheckingType = .no
+        searchBar.autocapitalizationType = .none
+        searchBar.autocorrectionType = .no
+        
+        return searchController
+    }
+    
+    private func getInitialData(map: AGSMap){
+        
+        // get the initial data for the map, this function should be called once the map is loaded
+        
+        guard let map = self.map else{
+            return
+        }
+        
+        // go through each layer and if it has a popup definition, editable and can create features
+        // then add it's info to the list
+        for layer in map.operationalLayers as NSArray as! [AGSLayer]{
+            
+            guard let fl = layer as? AGSFeatureLayer, let table = fl.featureTable as? AGSArcGISFeatureTable else {
+                continue
+            }
+            
+            fl.load{ [weak self] error in
+                
+                if error != nil{
+                    return
+                }
+                
+                guard let strongSelf = self else{
+                    return
+                }
+                
+                guard let popupDef = fl.popupDefinition, popupDef.allowEdit || table.canAddFeature else{
+                    return
+                }
+                
+                let tableTemplates = table.featureTemplates.map({
+                    FeatureTemplateInfo(featureLayer:fl, featureTable:table, featureTemplate:$0, swatch:nil)
+                })
+                
+                let typeTemplates = table.featureTypes
+                    .flatMap({ $0.templates })
+                    .map({ FeatureTemplateInfo(featureLayer:fl, featureTable:table, featureTemplate:$0, swatch:nil) })
+                
+                let infos = tableTemplates + typeTemplates
+                
+                // add to list of unfiltered infos
+                strongSelf.unfilteredInfos.append(contentsOf: infos)
+                
+                // re-assign to the current infos so we can update the tableview
+                // only should do this if not currently filtering
+                if !strongSelf.isFiltering{
```

Comment:
Could this be moved to a property observer of `unfilteredInfos`? That would work if you made `FeatureTemplateInfo` a struct.